beacon-chain/state/state-native/state_trie.go

```diff
@@ -0,0 +1,1060 @@
+		if b.rebuildTrie[field] {
+			err := b.resetFieldTrie(
+				field,
+				b.currentEpochAttestations,
+				fieldparams.CurrentEpochAttestationsLength,
+			)
+			if err != nil {
+				return [32]byte{}, err
+			}
+			delete(b.rebuildTrie, field)
+			return b.stateFieldLeaves[field].TrieRoot()
+		}
+		return b.recomputeFieldTrie(field, b.currentEpochAttestations)
+	case nativetypes.PreviousEpochParticipationBits:
+		return stateutil.ParticipationBitsRoot(b.previousEpochParticipation)
+	case nativetypes.CurrentEpochParticipationBits:
+		return stateutil.ParticipationBitsRoot(b.currentEpochParticipation)
+	case nativetypes.JustificationBits:
+		return bytesutil.ToBytes32(b.justificationBits), nil
+	case nativetypes.PreviousJustifiedCheckpoint:
+		return ssz.CheckpointRoot(hasher, b.previousJustifiedCheckpoint)
+	case nativetypes.CurrentJustifiedCheckpoint:
+		return ssz.CheckpointRoot(hasher, b.currentJustifiedCheckpoint)
+	case nativetypes.FinalizedCheckpoint:
+		return ssz.CheckpointRoot(hasher, b.finalizedCheckpoint)
+	case nativetypes.InactivityScores:
+		return stateutil.Uint64ListRootWithRegistryLimit(b.inactivityScores)
+	case nativetypes.CurrentSyncCommittee:
+		return stateutil.SyncCommitteeRoot(b.currentSyncCommittee)
+	case nativetypes.NextSyncCommittee:
+		return stateutil.SyncCommitteeRoot(b.nextSyncCommittee)
+	case nativetypes.LatestExecutionPayloadHeader:
+		return b.latestExecutionPayloadHeader.HashTreeRoot()
+	}
+	return [32]byte{}, errors.New("invalid field index provided")
+}
+
+func (b *BeaconState) recomputeFieldTrie(index nativetypes.FieldIndex, elements interface{}) ([32]byte, error) {
+	fTrie := b.stateFieldLeaves[index]
+	// We can't lock the trie directly because the trie's variable gets reassigned,
+	// and therefore we would call Unlock() on a different object.
+	fTrieMutex := fTrie.RWMutex
+	if fTrie.FieldReference().Refs() > 1 {
+		fTrieMutex.Lock()
+		fTrie.FieldReference().MinusRef()
+		newTrie := fTrie.CopyTrie()
+		b.stateFieldLeaves[index] = newTrie
+		fTrie = newTrie
+		fTrieMutex.Unlock()
+	}
+	// remove duplicate indexes
+	b.dirtyIndices[index] = slice.SetUint64(b.dirtyIndices[index])
+	// sort indexes again
+	sort.Slice(b.dirtyIndices[index], func(i int, j int) bool {
+		return b.dirtyIndices[index][i] < b.dirtyIndices[index][j]
+	})
+	root, err := fTrie.RecomputeTrie(b.dirtyIndices[index], elements)
+	if err != nil {
+		return [32]byte{}, err
+	}
+	b.dirtyIndices[index] = []uint64{}
+	return root, nil
+}
+
+func (b *BeaconState) resetFieldTrie(index nativetypes.FieldIndex, elements interface{}, length uint64) error {
+	fTrie, err := fieldtrie.NewFieldTrie(index, fieldMap[index], elements, length)
+	if err != nil {
+		return err
+	}
+	b.stateFieldLeaves[index] = fTrie
+	b.dirtyIndices[index] = []uint64{}
+	return nil
+}
+
+// ComputeFieldRootsWithHasher hashes the provided state and returns its respective field roots.
+func ComputeFieldRootsWithHasher(ctx context.Context, state *BeaconState) ([][]byte, error) {
```

Comment:
Im concerned about the complexity of this function. This is easily the biggest function in the codebase, and extremely hard to maintain. If we can split it into smaller functions it would be ideal. It has a ton of complexity around version checks as well, and ordering is super important...let's evaluate how to proceed on this